pkg/autometrics/ctx.go

```diff
@@ -0,0 +1,191 @@
+}
+
+// NewContext is a constructor taking the parent context as argument.
+//
+// It accepts 'nil' as the parent context. In this case the constructor
+// acts as if it received a new, fresh context.Background().
+func NewContext(parentCtx context.Context) Context {
+	if parentCtx == nil {
+		parentCtx = context.Background()
+	}
+	return Context{
+		TrackConcurrentCalls: true,
+		TrackCallerName:      true,
+		AlertConf:            nil,
+		Context:              parentCtx,
+	}
+}
+
+// SetTraceID sets the context's [TraceID]
+func (ctx *Context) SetTraceID(tid TraceID) {
+	ctx.Context = context.WithValue(ctx.Context, currentTraceId, tid)
+}
+
+// SetSpanID sets the context's [SpanID]
+func (ctx *Context) SetSpanID(sid SpanID) {
+	ctx.Context = context.WithValue(ctx.Context, currentSpanId, sid)
+}
+
+// SetParentSpanID sets the context's span's parent [SpanID]
+func (ctx *Context) SetParentSpanID(sid SpanID) {
+	ctx.Context = context.WithValue(ctx.Context, parentSpanId, sid)
+}
+
+// GetTraceID returns (_, false) if the context did not contain any trace id.
+func (c Context) GetTraceID() (TraceID, bool) {
+	if c.Context == nil {
+		return TraceID{}, false
+	}
+	tid, ok := c.Value(currentTraceId).(TraceID)
+	return tid, ok
+}
+
+// GetSpanID returns (_, false) if the context did not contain the current span id.
+func (c Context) GetSpanID() (SpanID, bool) {
+	if c.Context == nil {
+		return SpanID{}, false
+	}
+	sid, ok := c.Value(currentSpanId).(SpanID)
+	return sid, ok
+}
+
+// GetParentSpanID returns (_, false) if the context did not contain the parent's span id (including when we are in the root span).
+func (c Context) GetParentSpanID() (SpanID, bool) {
+	if c.Context == nil {
+		return SpanID{}, false
+	}
+	sid, ok := c.Value(parentSpanId).(SpanID)
+	return sid, ok
+}
+
+// FillTracingInfo ensures the context has a traceID and a spanID.
+// If they do not have this information, this method adds randomly
+// generated IDs in the context to be used later for exemplars
+//
+// The random generator is a PRNG, seeded with the timestamp of the first time new IDs are needed.
+func (c *Context) FillTracingInfo() {
+	// We are using a PRNG because FillTracingInfo is expected to be called in PreInstrument.
+	// Therefore it can have a noticeable impact on the performance of instrumented code.
+	// Pseudo randomness should be enough for our use cases, true randomness might introduce too much latency.
+	// randSource is initialized with a timestamp from the first time it is accessed in nanoseconds, which should
+	// be enough precision to avoid accidental collisions (imagine multiple services starting "at the same time" in a deployment).
+	if randSource == nil {
+		randSource = rand.New(rand.NewSource(time.Now().UnixNano()))
+	}
+
+	if parentSpanId, ok := c.GetSpanID(); ok {
+		c.SetParentSpanID(parentSpanId)
+	}
+
+	sid := SpanID{}
+	_, _ = randSource.Read(sid[:])
+	c.SetSpanID(sid)
+
+	// REVIEW: we might not want to fill the trace ID if it is absent.
```

Comment:
u checked this?

Reply:
yes, I should remove the comment. It seems that filling the trace ID with our own one is the correct thing to do if we didn't do it before. That's what most people would expect